Add GSoC 2026 blog post for extending Clang API Notes for C++ overloads (#89)
llvm/llvm-blog-www: content/posts/2026-08-21-gsoc-clang-api-notes-overloads.md

I was interested in this project because it connects Clang's semantic analysis with Swift interoperability. What made the project especially interesting to me was that a concise user-facing YAML extension had to be connected to Clang’s internal model of declarations, types, overloads, and source-level type spelling.

The goal of the project was to let API Notes target individual C++ function and method overloads. The work covered the public YAML design, parsing, binary serialization, declaration lookup, Sema integration, type normalization, diagnostics, and member-function object qualifiers.

## Background

API Notes allow additional information to be attached to declarations without modifying their original source headers. They are used in particular to refine how C, Objective-C, and C++ APIs are exposed to Swift.

Previously, an API Notes entry could identify a C++ function or method primarily by name. This is not sufficient when several overloads share that name:

```cpp
struct Widget {
  void setValue(int);
  void setValue(double);
};
```

> A name-only entry for `setValue` can apply to the whole overload set, but it cannot assign different annotations to `setValue(int)` and `setValue(double)`.

Reviewer: nit:

```suggestion
A name-only entry for `setValue` applies to the whole overload set, so it cannot assign different annotations to `setValue(int)` and `setValue(double)`.
```

The main challenge was therefore not only to select individual overloads, but to do so without changing the behavior of existing name-only API Notes files.

## Selecting C++ Overloads

> The design keeps the existing `Functions`, `Tags`, and `Methods` API Notes structure. The declaration name remains the primary lookup key, while an optional `Where` block narrows the candidate set.

Reviewer: "The design" is a bit abrupt since you haven't actually described it yet. Perhaps something like:

> API Notes are structured according to the kind of declaration they apply to, like `Functions`, `Tags`, and `Methods`. The C++ overload selectors I designed for my project extend the `Functions` and `Methods` schemas.

```yaml
Tags:
  - Name: Widget
    Methods:
      - Name: setValue
        Where:
          Parameters:
            - int
        SwiftName: setIntValue(_:)

      - Name: setValue
        Where:
          Parameters:
            - double
        SwiftName: setDoubleValue(_:)
```

This API Notes file uses the `Where.Parameters` selector to apply separate `SwiftName` annotations to the `int`- and `double`-parameter overloads of `setValue`.

> 

Reviewer: Suggestion: it might be illustrative to show the annotated example that applying these API Notes produces, i.e.,

```c++
struct Widget {
  SWIFT_NAME(setIntValue(_:)) void setValue(int);
  SWIFT_NAME(setDoubleValue(_:)) void setValue(double);
};
```

Here and elsewhere, at your discretion (adding this also increases verbosity, so don't overdo it either).

Author: Thanks! I added this after the first `Widget::setValue` example. I kept it to just that one place for now, since it seems like the clearest spot to show the effect without making the later examples too verbose.

```yaml
Functions:
  - Name: makeWidget
    Where:
      Parameters:
        - int
    SwiftName: makeWidgetFromInt(_:)

  - Name: makeWidget
    Where:
      Parameters:
        - double
    SwiftName: makeWidgetFromDouble(_:)
```

Name-only entries preserve their current behavior:

```yaml
Tags:
  - Name: Widget
    Methods:
      - Name: setValue
        Availability: nonswift
```

This entry still applies to every method named `setValue` in `Widget`. Omitting `Where.Parameters` leaves the parameter list unconstrained. Adding `Where.Parameters` narrows the entry to a specific explicit parameter list.

An explicitly empty parameter list has a different meaning:

```yaml
Methods:
  - Name: build
    Where:
      Parameters: []
    SwiftName: buildWithoutArguments()
```

`Where.Parameters: []` selects a declaration with no explicit source parameters. This differs from the name-only case above, which does not constrain parameters at all.

The parameter selector describes the complete explicit parameter list, so parameter count and order are significant.

## Matching C++ Types

Matching parameter types requires more than comparing raw strings.

> Consider a declaration that uses an alias:

Reviewer: ```suggestion
Consider a method whose parameter uses a type alias:
```

Selector-only nullability is stripped during selector formation because API Notes can add or replace nullability independently. This includes nested nullability in pointer, array, and function-pointer type components.

Top-level `const` and `volatile` also need special handling:

```cpp
void process(int);
void process(const int);
```

These do not declare different C++ overloads. The selector representation therefore removes top-level qualifiers in cases where they do not contribute to overload identity, including the supported by-value and top-level pointer cases.

Because top-level `const` is stripped in this position, the second `process` declaration is treated as a redeclaration of the first rather than as a separate overload.

The goal is not to implement complete semantic type equivalence. Instead, the matching policy preserves useful source-level distinctions while normalizing differences that should not select separate overloads.

## Matching the Implicit Object

Explicit parameters are not enough to distinguish every C++ method overload.

> Member functions can differ through qualifiers on their implicit object parameter:

Reviewer: nit: these could probably be one paragraph.

```suggestion
Explicit parameters are not enough to distinguish every C++ method overload. Member functions can differ through qualifiers on their implicit object parameter:
```

```cpp
struct Builder {
  Result build() &;
  Result build() &&;

>   Result build() const &;

Reviewer: I wonder whether you should first show the more common case of a method having just `const` and non-`const` overloads first, the archetypical example being `operator[]` (I think support for naming operators landed [recently](https://github.com/llvm/llvm-project/pull/177382)).

Overloading via l- and r-value-this is less common; maybe show this later as a secondary, "advanced" example.

Author: Good point, thanks! I changed this section to lead with the `operator[]` const/non-const example using `Where.Object`/`Object.Const`, and kept the `Builder::build() &/&&` case afterward as the secondary ref-qualified example.

};
```

> All three methods have the same name and no explicit parameters. `Where.Parameters: []` alone cannot distinguish them.

Reviewer: ```suggestion
All three methods have the same name and no explicit parameters, so `Where.Parameters: []` alone cannot distinguish them.
```

> The selector model therefore includes an `Object` constraint:

Reviewer: nit: phrasing

```suggestion
Method selectors can therefore also include an `Object` constraint:
```

In particular, this makes it explicit that `Object` constraints apply specifically to methods.

## Implementation and Results

Supporting overload-aware API Notes required changes throughout the existing pipeline:

```text
YAML
  -> API Notes model
  -> binary serialization
  -> declaration lookup
  -> overload-specific filtering
  -> Sema applies API Notes effects
```

The implementation was split into focused patches.

The first changes added [YAML parsing and data-model support](https://github.com/llvm/llvm-project/pull/203227) so that multiple entries with the same declaration name but different selectors could be represented.

The next part extended the [binary API Notes format](https://github.com/llvm/llvm-project/pull/204147). Overload-specific entries must remain distinct when API Notes are compiled and later loaded again. The serialization also preserves the difference between an omitted parameter constraint and an explicitly empty parameter list.

The [Sema integration](https://github.com/llvm/llvm-project/pull/205307) first finds API Notes by declaration context and name, then uses the declaration’s explicit parameter types to select the overload-specific entry and apply its effects.

> Legacy name-only lookup remains unchanged. An explicit selector, however, does not silently fall back to an unrelated broad match when it fails. Such a fallback could apply an annotation intended for one overload to another declaration with the same name.

Reviewer: IMO I don't think you need to mention what the explicit selector _doesn't_ do (especially since you already discussed this in the design section). It makes the writing less clear, and is not important here because this is something you _didn't_ do.

You can just elide the "Legacy name-only lookup remains unchanged." into the previous paragraph.

Additional work adds [diagnostics for malformed, duplicate, and unmatched selectors](https://github.com/llvm/llvm-project/pull/209408). The [normalization patch](https://github.com/llvm/llvm-project/pull/213043) refines selector lookup for aliases, nullability, qualifiers, and supported spelling differences, while the [object-qualifier patch](https://github.com/llvm/llvm-project/pull/216148) adds `Where.Object` matching. A small [diagnostics follow-up](https://github.com/llvm/llvm-project/pull/216272) makes generic `-Wapinotes` warnings visible for system headers, which matters because API Notes are commonly attached to SDK headers. Tests cover the parser, serialization, Sema lookup, normalization behavior, aliases, default arguments, static methods, zero-parameter declarations, and object-qualified member functions.

The result is an overload-aware selector model that supports both global functions and C++ methods while preserving existing API Notes behavior.

## Future Template Design

> The future template design work explored how the overload selector model might eventually extend from concrete functions and methods to C++ templates. Template matching was not implemented in the current patches, but the design work helped identify which parts of the selector model should remain extensible.

Reviewer: Phrasing nit:

```suggestion
The overload selector model only applies to concrete methods and functions. As part of my project, I also explored how the overload selector model could be extended to support C++ templates in the future. Function template matching is not currently implemented, but this design work helped identify which parts of the selector model should remain extensible.
```

Selector-only nullability is stripped during selector formation because API Notes can add or replace nullability independently. This includes nested nullability in pointer, array, and function-pointer type components.

Top-level `const` and `volatile` also need special handling:

```cpp
void process(int);
void process(const int);
```

These do not declare different C++ overloads. The selector representation therefore removes top-level qualifiers in cases where they do not contribute to overload identity, including the supported by-value and top-level pointer cases.

Because top-level `const` is stripped in this position, the second `process` declaration is treated as a redeclaration of the first rather than as a separate overload.

The goal is not to implement complete semantic type equivalence. Instead, the matching policy preserves useful source-level distinctions while normalizing differences that should not select separate overloads.

## Matching the Implicit Object

> Explicit parameters are not enough to distinguish every C++ method overload. Member functions can differ through qualifiers on their implicit object parameter. A common example is `operator[]`, the subscript operator, with separate mutable and read-only overloads:

Reviewer: nit: I think you should explicitly spell out `const` vs non-`const` here, which are more precise than "mutable" and "read-only" in the context of describing C++:

```suggestion
Explicit parameters are not enough to distinguish every C++ method overload. Member functions can differ through qualifiers on their implicit object parameter. A common example is `operator[]`, the subscript operator, with separate `const` and non-`const` overloads:
```

```cpp
struct Buffer {
  Element &operator[](int);
  const Element &operator[](int) const;
};
```

Both overloads have the same name and explicit parameter list. `Where.Parameters` can identify the `int` parameter, but it cannot distinguish the non-`const` overload from the `const` overload by itself.

The selector model therefore includes an `Object` constraint:

```yaml
Tags:
  - Name: Buffer
    Methods:
      - Name: operator[]
        Where:
          Parameters:
            - int
          Object:
            Const: false
        SwiftName: mutableElement(at:)

      - Name: operator[]
        Where:
          Parameters:
            - int
          Object:
            Const: true
        SwiftName: element(at:)
```

The `Object` selector describes `const`, `volatile`, and reference qualification on the implicit C++ object.

The same model can also distinguish lvalue- and rvalue-qualified methods:

```cpp
struct Builder {
  Result build() &;
  Result build() &&;
  Result build() const &;
};
```

> For these overloads, `Where.Parameters: []` identifies the empty explicit parameter list, while `Where.Object.Ref` distinguishes lvalue and rvalue receivers.

Reviewer: I wonder if you should show the API Notes file for this?

Author: That makes sense, thanks. I added a compact API Notes block for the `Builder` example showing how `Where.Parameters: []` combines with `Object.Ref`/`Object.Const` for these overloads.
